src/python/tests/appengine/libs/query/datastore_query_test.py

```diff
@@ -171,6 +171,24 @@ def test_greater_or_equal(self):
     self.assertEqual(total_count, 5)
     self.assertFalse(has_more)
 
+  def test_negative_page(self):
+    """Test getting a negative page."""
+    query = datastore_query.Query(TestDatastoreModel)
+    query.filter_in('tokens', ['a', 'b'])
+    query.filter('boolean_value', True)
+    query.order('datetime_value', is_desc=True)
+
+    items, total_pages, total_items, has_more = query.fetch_page(
+        page=-5, page_size=2, projection=['tokens'], more_limit=4)
+    self.assertListEqual([self.mocks[0].key.id(), self.mocks[2].key.id()],
+                         [item.key.id() for item in items])
+    self.assertListEqual([['a'], ['a']], [item.tokens for item in items])
+    with self.assertRaises(ndb.UnprojectedPropertyError):
+      _ = [item.boolean_value for item in items]
+    self.assertEqual(6, total_items)
```

Comment:
Compared to test_third_page, why did the total_items and total_pages decrease here ?

Reply:
The page we are loading here is 1st which is followed by 2 pages(2nd and 3rd) each containing 2 items each. So total_items becomes 6 and total_pages becomes 3. 

In the 3rd page test case, 2 pages are before the 3rd one(1st and 2nd) and 2 are followed by it(4th and 5th) which results in total_items and total_pages as 10 and 5 respectively.

Reply:
How do their total items differ if the query is the same. that is what confused me.

Reply:
Oh, actually this wrapper doesn't returns the total_items that are present in the datastore, this just returns how many are there in the particular page limit we are looking at. Such as if the if we are at page 10 and it has 2 pages before this in display and 2 after and each page have 20 items. Then total_items will return 5*20=100. 

In the above test case of 3rd page you can also see it is 10 but in the datastore prepration, it actually adds 30 entries to the datastore.

Reply:
ah great, thanks for explaining!

Reply:
no problem :)